packages/devtools_app/lib/src/charts/chart.dart

```diff
@@ -0,0 +1,588 @@
+    Canvas canvas,
+    PaintCharacteristics characteristics,
+    double x,
+    double y,
+  ) {
+    final paint = Paint()
+      ..style = characteristics.symbol == ChartSymbol.disc
+          ? PaintingStyle.fill
+          : PaintingStyle.stroke
+      ..strokeWidth = characteristics.strokeWidth
+      ..color = characteristics.color;
+
+    canvas.drawCircle(Offset(x, y), characteristics.diameter, paint);
+  }
+
+  // TODO(terry): Use bezier path.
+  void drawDashed(Canvas canvas, PaintCharacteristics characteristics, double x,
+      double y, double tickWidth) {
+    assert(characteristics.symbol == ChartSymbol.dashedLine);
+    drawLine(
+      canvas,
+      characteristics,
+      x,
+      y,
+      tickWidth,
+    );
+  }
+
+  void drawConnectedLine(
+    Canvas canvas,
+    PaintCharacteristics characteristics,
+    double startX,
+    double startY,
+    double endX,
+    double endY,
+  ) {
+    final paint = Paint()
+      ..style = PaintingStyle.stroke
+      ..strokeWidth = characteristics.strokeWidth
+      ..color = characteristics.color;
+
+    canvas.drawLine(Offset(startX, startY), Offset(endX, endY), paint);
+  }
+
+  void drawLine(
+    Canvas canvas,
+    PaintCharacteristics characteristics,
+    double x,
+    double y,
+    double tickWidth,
+  ) {
+    final paint = Paint()
+      ..style = PaintingStyle.stroke
+      ..strokeWidth = characteristics.strokeWidth
+      ..color = characteristics.color;
+
+    canvas.drawLine(Offset(x, y), Offset(x + tickWidth, y), paint);
+  }
+
+  void drawFillArea(
+    Canvas canvas,
+    PaintCharacteristics characteristics,
+    double x0,
+    double y0,
+    double y0Bottom,
+    double x1,
+    double y1,
+    double y1Bottom,
+  ) {
+    final paint = Paint()
+      ..style = PaintingStyle.fill
+      ..strokeWidth = characteristics.strokeWidth
+      ..color = characteristics.color.withAlpha(100);
+
+    final fillArea = Path()
+      ..moveTo(x0, y0Bottom)
+      ..lineTo(x0, y0)
+      ..lineTo(x1, y1)
+      ..lineTo(x1, y1Bottom)
+      ..lineTo(x0, y0Bottom);
+    fillArea.close();
+    canvas.drawPath(fillArea, paint);
+
+    fillArea.reset();
+  }
+
+  @override
+  bool shouldRepaint(ChartPainter oldDelegate) =>
+      true; // percentage != oldDelegate.percentage;
+
+  /// Return the largest Y value in a particular trace if traceIndex is passed or
+  /// all traces if traceIndex is not passed in.
+  double maxValue({int traceIndex}) {
+    var maxValue = 0.0;
+    if (traceIndex == null) {
+      for (var index = 0; index < chartController.traces.length; index++) {
+        final trace = chartController.traces[index];
+        final traceMax =
+            trace.data.reduce((curr, next) => curr.y > next.y ? curr : next).y;
+        if (traceMax > maxValue) maxValue = traceMax;
+      }
+    } else {
+      final trace = chartController.traces[traceIndex];
+      maxValue =
+          trace.data.reduce((curr, next) => curr.y > next.y ? curr : next).y;
```

Comment:
add a helper to compute the max value from a trace.

Reply:
Done.